### ISSUE TITLE
ANTLR: Deciding on how far to go with this

### ISSUE THREAD

**Issue by RexJaeschke** [dotnet/csharpstandard#37]
On 2020-10-03, I circulated the following paper: [ANTLR and the C# Spec.docx](https://github.com/ECMA-TC49-TG2/csharpstandard/files/5575949/ANTLR.and.the.C.Spec.docx). In that, I discussed using the ANTLR grammar notation and the non-trivial amount of work I thought it would take to get the grammar to validate using the ANTLR toolkit.

Neal has reviewed that paper and his feedback follows in a comment below, with a proposal on how we might go forward. Once we discuss (and probably agree with) his suggestion, I'll update this comment by adding the remaining questions and to keep track of our decisions on those questions .

Here are the ANTLR-related questions and their status:

Proposal # | Status | Description
------------ | ------- | -------------
1 | Yes it should | Decide whether our grammar should be complete enough to validate (3 Rex's paper). **Note that *validation* here means "get through the ANTLR tool without error," which doesn't necessarily mean that any parser generated will be complete/correct.** This is especially true for semantic predicates.
2 | Resolved, PRs #208, #209  | Change case of lexer rule names (5.1 Rex's paper)
3 | Resolved, PR #261 | Expanding lexer rule alternatives of the form <...> (3 Rex's paper)
4 | Resolved, PR #225 | Limited use of token names (5.3 Rex's paper)
5 | In-progress | Resolve Mutual Left-Recursion issue (15 Rex's paper)
6 | Resolved, PR #233 | How to handle the unsafe extensions
7 | Resolved | Marking lexer helper rules with `fragment` 
8 | Resolved | Ultimately, revise 7.2, "Grammars" in lexical-structure.md to say what we do/don't do w.r.t full grammar validation

**When you add a comment to this issue, please limit it to a single Proposal # and say which Proposal # that is.**

Any decisions we make that require changes to the grammar notation will not only have to be made to the v5 grammar we're starting with in v6, but also possibly to the v6 (and v7) PRs in waiting.

**Comment by Nigel-Ecma** [comment]
I've taken a brief look at ANTLR and I beleive the rules Rex quotes above can be written:

```
Not_Slash_Or_Asterisk
    : ~[*/] // Any Unicode character except / or *
    ;

Letter_Character
    : [\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Lo}\p{Nl}] // A Unicode Character of classes Lu, Ll, Lt, Lm, Lo, or Nl
    | Unicode_Escape_Sequence { IsLetterCharacter() }? // as previous but using an escape sequence
    ;
```

The last rule includes a *semantic predicate*, `{ IsLetterCharacter() }?` this a way context-sensitive or other semantic checks can be included in the syntactical grammar. If the function `IsLetterCharacter()` returns false the alternative is "unviable". Semantic predicates can be included in both parser and lexer rules. The predicates themselves can be defined in the same file as the grammar; are written in Java, C++, Javascript or any other language ANTLR supports; and have access to the source being parsed/lexed.

The above doesn't tell us that all the content-sensitive/semantic needs of the C# grammar can be supported, but I think we should find out – I have a certain unease in publishing a language Standard with a **known** invalid/incomplete grammar.

**Comment by RexJaeschke** [comment]
Regarding Proposal 3, "<...>" (https://github.com/dotnet/csharpstandard/issues/37#issuecomment-789755202), I've taken into account input from Nigel (https://github.com/dotnet/csharpstandard/issues/37#issuecomment-793154408) and Sam (https://github.com/dotnet/csharpstandard/issues/37#issuecomment-796165854).

The handling of the Unicode sequences in identifiers results in the following grammar, with current alternatives to be replaced commented out and followed by their replacement(s):

```ANTLR
Letter_Character
//    : '<A Unicode character of classes Lu, Ll, Lt, Lm, Lo, or Nl>'
//    | '<A Unicode_Escape_Sequence representing a character of classes Lu, Ll, Lt, Lm, Lo, or Nl>'

    : [\p{L}\p{Nl}]     // category Letter, all subcategories; category Number, subcategory letter
    | Unicode_Escape_Sequence { IsLetterCharacter() }?
    ;

Combining_Character
//    : '<A Unicode character of classes Mn or Mc>'
//    | '<A Unicode_Escape_Sequence representing a character of classes Mn or Mc>'

    : [\p{Mn}\p{Mc}]    // category Mark, subcategories non-spacing and spacing combining
    | Unicode_Escape_Sequence { IsCombiningCharacter() }?
    ;

Decimal_Digit_Character
//    : '<A Unicode character of the class Nd>'
//    | '<A Unicode_Escape_Sequence representing a character of the class Nd>'

    : [\p{Nd}]      // category Number, subcategory decimal digit
    | Unicode_Escape_Sequence { IsDecimalDigitCharacter() }?
    ;

Connecting_Character
//    : '<A Unicode character of the class Pc>'
//    | '<A Unicode_Escape_Sequence representing a character of the class Pc>'

    : [\p{Pc}]      // category Punctuation, subcategory connector
    | Unicode_Escape_Sequence { IsConnectingCharacter() }?
    ;

Formatting_Character
//    : '<A Unicode character of the class Cf>'
//    | '<A Unicode_Escape_Sequence representing a character of the class Cf>'

    : [\p{Cf}]      // category Other, subcategory format
    | Unicode_Escape_Sequence { IsFormattingCharacter() }?
    ;
```

While the definitions of the `Is*` predicate methods need to be in the grammar file input to ANTLR when validation occurs (as part of Bill's actions), I see a way to smuggle them in *without* having to include them in the spec.

This leaves the final two uses of the `'<...>' notation

```ANTLR
Available_Identifier
//    : '<An Identifier_Or_Keyword that is not a Keyword>'

    : Identifier_Or_Keyword { IsNotAKeyword() }?
    ;

Conditional_Symbol
//    : '<Any Identifier_Or_Keyword except true or false>'

    : Identifier_Or_Keyword { IsNotTrueOrFalse() }?
    ;
```

Nigel and I support the use of semantic predicates here as well, which I have shown above. Sam proposed some rearrangement of the grammar instead. @sharwell, is the semantic predicate approach OK with you?

How does this all look to you, @Nigel-Ecma?

**Comment by RexJaeschke** [comment]
## Proposal 5: Resolving Mutual Left-Recursion issue

@MadsTorgersen, @sharwell, @Nigel-Ecma

There are three lexer rules that have this problem, all in the preprocessor grammar, w.r.t *Pp_Expression*s involving `||`, `&&`, `==`, and `!=`.

On a whim, I rewrote those rules, as follows, with each commented-out line being replaced by the one following, as follows:

```ANTLR
Pp_Or_Expression
    : Pp_And_Expression

//  | Pp_Or_Expression Whitespace? '||' Whitespace? Pp_And_Expression
    | Whitespace? Pp_And_Expression '||' Pp_Or_Expression Whitespace?
    ;
    
Pp_And_Expression
    : Pp_Equality_Expression

//  | Pp_And_Expression Whitespace? '&&' Whitespace? Pp_Equality_Expression
    | Whitespace? Pp_Equality_Expression '&&' Pp_And_Expression Whitespace?
    ;

Pp_Equality_Expression
    : Pp_Unary_Expression

//  | Pp_Equality_Expression Whitespace? '==' Whitespace? Pp_Unary_Expression
    | Whitespace? Pp_Unary_Expression '==' Pp_Equality_Expression Whitespace? 
    
//  | Pp_Equality_Expression Whitespace? '!=' Whitespace? Pp_Unary_Expression
    | Whitespace? Pp_Unary_Expression '!=' Pp_Equality_Expression Whitespace?  
    ;
```

All I did was simply swap the two operands over, and it validated! But are the new forms equivalent/correct?

**Comment by Nigel-Ecma** [comment]
> All I did was simply swap the two operands over, and it validated! But are the new forms equivalent/correct?
> 
It changes the grouping/associativity, e.g. given a || b || c one version parses it is (a || b) || c the other as a || (b || c)

Does it matter *in this case* (obviously it can matter in some cases)?

(Sorry no answer right now, I’m distracting myself clearing email when I should be doing something else which I must get back to…)

**Comment by RexJaeschke** [comment]
This Issue was closed automatically when the unsafe grammar PR was merged, but we still have the left-recursive situation to deal with, so am re-opening.

**Comment by Nigel-Ecma** [comment]
Issues #259 & #260 raise questions which are dependent on this issue and investigations into the grammar it has triggered. Those questions should be considered in any resolution of this one.

**Comment by RexJaeschke** [comment]
## Proposal 7: Marking lexer helper rules with `fragment`

Previously, it seemed that applying `fragment` to a lexer rule might be useful, but not necessary. However, as Nigel discovered, it really should be used on all lexer rules that help other rules. That is, any lexer rule that does *not* result in a complete token should be marked `fragment`. For example:

```ANTLR
Decimal_Integer_Literal     // forms a complete token; is not a helper rule
    : Decimal_Digit+ Integer_Type_Suffix?
    ;
    
fragment
Decimal_Digit         // helper rule; should have fragment prefix
    : '0'..'9'
    ;
    
fragment
Integer_Type_Suffix         // helper rule; should have fragment prefix
    : 'U' | 'u' | 'L' | 'l' | 'UL' | 'Ul' | 'uL' | 'ul' | 'LU' | 'Lu' | 'lU' | 'lu'
    ; 
```

As such, Nigel will add this prefix, as appropriate, as part of his efforts.

**Comment by RexJaeschke** [comment]
## Proposal 8: Updating text in lexical-structure.md to state our intent w.r.t using ANTLR notation vs. having an ANTLR-ready grammar.

Nigel is handling this as part of his grammar overhaul PRs.

**Comment by RexJaeschke** [comment]
@Nigel-Ecma I'm inclined to close this Issue, unless you want to keep it open as a hook for things you are still working on.

**Comment by Nigel-Ecma** [comment]
I think closing it is fine, you could reference the various PR’s now merged (interpolated strings mentioned in #37 are still in progress but a PR will come)

**Comment by RexJaeschke** [comment]
Almost all of the outstanding tasks were handled by Nigel in various PRs, including #339, #342, and #351. Closing this Issue with Nigel's consent.
